### PR TITLE
tomlplusplus/2.5.0: changed minimal compiler version

### PR DESCRIPTION
**Pull request by YaZasnyal**
tomlplusplus/2.5.0

Current recipe requires MSVC 16 compiler but 15.3 should be enough ([link](https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-160#c-standards-support)). 
Or maybe we can remove compiler checking completely because ```tools.check_min_cppstd``` as I understand checks cpp standard support.
 
---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
> maybe we can remove compiler checking completely because tools.check_min_cppstd as I understand checks cpp standard support.

No, because it requires C++17. 

> Current recipe requires MSVC 16 compiler but 15.3 should be enough (link).

Depending the required feature, yes. Let's see if it pass by the CI

**Review by dvetutnev — APPROVED** [review]
Yes, MSVC 15.3 support C++17.

**Comment by YaZasnyal** [comment]
As I can see all MSVC builds have passed successfully.

As for other failed builds I have figured out that version 2.0.0 had a problem with include guards that was fixed later in 2.1.0 ([commit](https://github.com/marzer/tomlplusplus/commit/35ada5d851f57a8150720b77e67cb6f24e9ef0ef)). This causes error during test compilation. There are multiple solutions to this:
* Add additional condition to conanfile.py of the test package that checks for version 2.0.0 exactly and adds ```TOMLPP_BUILD_SINGLE_ONLY```. This solution is not ideal because corner cases are bad;
* We can switch order of the includes. The single-header version will be included after include guard definition. This is bad because multiple header version is deprecated and we will test only this deprecated version of the library;
* We can make additional compilation unit that will test multiple-header version. This will check both versions of the library but won't check them together and such errors will remain hidden;
* We can exclude version 2.0.0 whatsoever because it is faulty.

I think that the exclusion of the 2.0.0 version is the most correct solution because it does not require to add checks for this exact version to the test package and prevents end users of using faulty packages. If you think that other solution should be chosen feel free to contact me and I will add necessary changes to the PR.

As for MacOS builds. I am not able to debug MacOS builds because I don't have any.

**Comment by YaZasnyal** [comment]
@uilianries Can you comment on [this](https://github.com/conan-io/conan-center-index/pull/7350#issuecomment-932928222)? Also I have tried to build on mac and was unable to reproduce behavior in CI =(

**Comment by YaZasnyal** [comment]
I have created a separate test for multiple header version.

Also I don't understand how this tests passed previously because there are definitely some bugs in earlier version of the library. This is why I had to add some ugly conditionals to the ```conanfile.txt```.

MSVC 15 support was added in 2.2.0.
https://github.com/marzer/tomlplusplus/releases/tag/v2.2.0

MSVC 16 has a bug in exactly 2.1.0
https://github.com/marzer/tomlplusplus/issues/56

apple-clang bugs were first fixed in 2.3.0
https://github.com/marzer/tomlplusplus/releases/tag/v2.3.0

File: recipes/tomlplusplus/all/conanfile.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def _minimum_cpp_standard(self):
     @property
     def _minimum_compilers_version(self):
         return {
-            "Visual Studio": "16",
+            "Visual Studio": "15",
```
Comment (prince-chrismc) [inline recipes/tomlplusplus/all/conanfile.py:30]:
```suggestion
            "Visual Studio": "16" if tools.Version(self.version) < "2.2.0" else "15",
```

File: recipes/tomlplusplus/all/conanfile.py [hunk 2]
```diff
@@ -49,6 +49,18 @@ def configure(self):
...
+            raise ConanInvalidConfiguration("The current compiler {} {} is supported in version >= 2.3.0".format(
+                    self.settings.compiler, self.settings.compiler.version))
+        
+        if self.settings.compiler == "Visual Studio":
+            if tools.Version(self.version) == "2.1.0":
+                raise ConanInvalidConfiguration("The current compiler {} {} is unable to build version 2.1.0".format(
+                        self.settings.compiler, self.settings.compiler.version))
+            if tools.Version(self.settings.compiler.version) == "15" and tools.Version(self.version) < "2.2.0":
+                raise ConanInvalidConfiguration("The current compiler {} {} is supported in version >=2.2.0".format(
+                        self.settings.compiler, self.settings.compiler.version))
```
Comment (prince-chrismc) [inline recipes/tomlplusplus/all/conanfile.py:62]:
```suggestion
```

**Review by prince-chrismc — COMMENTED** [review]
Small simplification

**Review by prince-chrismc — COMMENTED** [review]
There's a few other "modernize"-ing things for the recipe (sorry to add on an old PR)

If you could please change

```diff
-def configure(self):
+def validate(self);
```

As well as replace the `source` method with https://github.com/conan-io/conan-center-index/blob/36f1df6d7aa464d0109088aed8587fc5b20529f5/recipes/ade/all/conanfile.py#L45-L46

It will help update this recipe :smile:

**Comment by prince-chrismc** [comment]
Other than my comments, I approve.

Last time changes were made header only was test on a single platform so the version bumps did not find the compilation errors (or lack of support) so the changes seem to make sense.

**Review by prince-chrismc — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
